### ISSUE TITLE
Thread-safety concern related to interpolation parsing

### ISSUE THREAD

**Issue by odelalleau** [omry/omegaconf#715]
**Describe the bug**

It occurred to me that we may want to ensure that OmegaConf is thread-safe (at least on simultaneous reads). But the [cache mechanism](https://github.com/omry/omegaconf/blob/2186b1da6134129c9f4ca89e4b479be445a2690d/omegaconf/grammar_parser.py#L121) I introduced in #445 to speed-up the parsing of interpolations is not thread-safe.

**Expected behavior**

Parsing interpolations in parallel from multiple threads should not fail because of the cache mentioned above.

**Comment by omry** [comment]
Thread safety is not a current goal.
We have several cache mechanisms besides this one that are all potentially unsafe. So far this has not came up in user reports.

**Comment by odelalleau** [comment]
Yeah I assumed it wasn't a top priority, but I also thought it wouldn't hurt to address this specific one.

For what it's worth I ended up thinking about it because I realized my own project was multi-threaded and might run into problems because of this specific issue (although it's unlikely to happen because my reads are infrequent enough). The tricky thing is that even if it happens, I may not notice it because it might not crash (one potential side effect would be reading an incorrect config value, but in my specific use case that would essentially give incorrect results).

I won't push for it though if you feel like it's not worth it -- I'll probably duplicate my config instead in my project anyway just to be 100% safe :)
